### PR TITLE
feat/swarinfo 219

### PR DESCRIPTION
**Pull request by t-constantin**
Huge PR containing major overhaul of the application.

Instead of different activities, we only have one.
In this activity, a NavigationController changes to UI without changing the logic behind it.
It is then possible to make the communication logic persistent throughout the application.
Each fragment can contain it's own logic afterwards or be used only as a view.

Navigation Bottom Bar contains the different views.
The Connection tab shows (as a notification) the state of the current communication connection

Simplified the "onCreate" method on the main Activity

### CODE REVIEW

**Review by stlaperriere — APPROVED** [review]
Love how the code was reorganised to allow dynamic switching between different fragments. Looks clean and very scalable! GJ 👍

File: HiveAR/app/src/main/java/com/swarmus/hivear/activities/MainActivity.java [hunk 1]
```diff
@@ -1,37 +1,272 @@
...
+
+        final Observer<Integer> portObserver = p -> tcpDevice.setServerPort(p);
+        tcpSettingsViewModel.getPort().observe(this, portObserver);
+    }
+
+    private void setUpSerialCommunication() {
+        serialDevice = new SerialDevice();
+        serialDevice.init(this, connectionCallback);
+
+        IntentFilter filter = new IntentFilter(SerialDevice.ACTION_SERIAL_DEVICE_CHANGED);
```
Comment (t-constantin) [inline HiveAR/app/src/main/java/com/swarmus/hivear/activities/MainActivity.java:173]: Must be before line 171 in order to list device already connected when launching app.

File: HiveAR/app/src/main/java/com/swarmus/hivear/activities/MainActivity.java [hunk 2]
```diff
@@ -1,37 +1,272 @@
...
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
 
 public class MainActivity extends AppCompatActivity {
+    BottomNavigationView bottomNavigationView;
+    NavHostFragment navHostFragment;
+
+    private SerialDevice serialDevice;
```
Comment (t-constantin) [inline HiveAR/app/src/main/java/com/swarmus/hivear/activities/MainActivity.java:47]: Change both devices to CommunicationDevice

File: HiveAR/app/src/main/java/com/swarmus/hivear/activities/MainActivity.java (hunk 1 above)
Comment (t-constantin) [reply HiveAR/app/src/main/java/com/swarmus/hivear/activities/MainActivity.java:173]: Remove this receiver and simply use viewmodel directly inside of serial device

File: HiveAR/app/src/main/java/com/swarmus/hivear/activities/MainActivity.java [hunk 3]
```diff
@@ -1,37 +1,272 @@
...
+        setUpSerialCommunication();
+
+        currentCommunicationDevice = serialDevice;
+    }
+
+    private void setUpTCPCommunication() {
+        tcpDevice = new TCPDevice(DEFAULT_IP_ADDRESS, DEFAULT_PORT);
+        tcpDevice.init(this, connectionCallback);
+
+        TcpSettingsViewModel tcpSettingsViewModel = new ViewModelProvider(this).get(TcpSettingsViewModel.class);
```
Comment (t-constantin) [inline HiveAR/app/src/main/java/com/swarmus/hivear/activities/MainActivity.java:161]: remove and read viewmodel from tcpdevice

File: HiveAR/app/src/main/java/com/swarmus/hivear/fragments/TcpSettingsFragment.java [hunk 4]
```diff
@@ -53,15 +60,14 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
...
-            ipInputEditText.setOnFocusChangeListener(new View.OnFocusChangeListener() {
-                @Override
-                public void onFocusChange(View v, boolean hasFocus) {
-                    if (!hasFocus) {
-                        tcpSettingsViewModel.getIpAddress().setValue(Objects.requireNonNull(ipInputEditText.getText()).toString());
-                        ipInputEditText.clearFocus();
-                    }
+            ipInputEditText.setOnFocusChangeListener((v, hasFocus) -> {
+                if (!hasFocus) {
+                    tcpSettingsViewModel.getIpAddress().setValue(Objects.requireNonNull(ipInputEditText.getText()).toString());
```
Comment (t-constantin) [inline HiveAR/app/src/main/java/com/swarmus/hivear/fragments/TcpSettingsFragment.java:69]: addtextchsngedlistener instead

**Pull request merged**